src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -127801,6 +127801,4340 @@ static T F2<T>()
+                // (6,24): error CS0453: The type 'T' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'Nullable<T>'
+                //     public override T? F3<T>() => default;
+                Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "F3").WithArguments("System.Nullable<T>", "T", "T").WithLocation(6, 24),
+                // (8,24): error CS0508: 'B.F5<T>()': return type must be 'T' to match overridden member 'A.F5<T>()'
+                //     public override T? F5<T>() => default;
+                Diagnostic(ErrorCode.ERR_CantChangeReturnTypeOnOverride, "F5").WithArguments("B.F5<T>()", "A.F5<T>()", "T").WithLocation(8, 24),
+                // (8,24): error CS0453: The type 'T' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'Nullable<T>'
+                //     public override T? F5<T>() => default;
+                Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "F5").WithArguments("System.Nullable<T>", "T", "T").WithLocation(8, 24),
+                // (10,24): error CS0508: 'B.F7<T>()': return type must be 'T' to match overridden member 'A.F7<T>()'
+                //     public override T? F7<T>() => default;
+                Diagnostic(ErrorCode.ERR_CantChangeReturnTypeOnOverride, "F7").WithArguments("B.F7<T>()", "A.F7<T>()", "T").WithLocation(10, 24),
+                // (10,24): error CS0453: The type 'T' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'Nullable<T>'
+                //     public override T? F7<T>() => default;
+                Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "F7").WithArguments("System.Nullable<T>", "T", "T").WithLocation(10, 24),
+                // (11,24): error CS0508: 'B.F8<T>()': return type must be 'T' to match overridden member 'A.F8<T>()'
+                //     public override T? F8<T>() => default;
+                Diagnostic(ErrorCode.ERR_CantChangeReturnTypeOnOverride, "F8").WithArguments("B.F8<T>()", "A.F8<T>()", "T").WithLocation(11, 24),
+                // (11,24): error CS0453: The type 'T' must be a non-nullable value type in order to use it as parameter 'T' in the generic type or method 'Nullable<T>'
+                //     public override T? F8<T>() => default;
+                Diagnostic(ErrorCode.ERR_ValConstraintNotSatisfied, "F8").WithArguments("System.Nullable<T>", "T", "T").WithLocation(11, 24));
+
+            var sourceB2 =
+@"#nullable enable
+class B : A
+{
+    public override T? F1<T>() where T : default => default;
+    public override T? F2<T>() where T : default => default;
+    public override T? F3<T>() where T : default => default;
+    public override T? F4<T>() where T : default => default;
+    public override T? F5<T>() where T : default => default;
```

Comment:
Was a bit surprised that this override worked. I thought `where T: default` was only used in the case where it was the default. Here it is overriding `where T: notnull`. Are there other cases where this is allowed (besides interfaces)

Reply:
`where T : default` is allowed when the base uses `where T : notnull` because `notnull` does not constrain the type parameter to a reference type or value type. The `default` constraint is allowed unless the type parameter is explicitly constrained to `class`, `struct`, `unmanaged`, or an explicit reference or value type. As you said, that means the `default` constraint is allowed when the type parameter is only constrained to an interface.

---
In reply to: [457840762](https://github.com/dotnet/roslyn/pull/45993#discussion_r457840762) [](ancestors = 457840762)